js/private/js_image_layer.bzl

```diff
@@ -0,0 +1,182 @@
+"""Rules for creating container image layers from js_binary targets
+
+For example, this js_image_layer target outputs `node_modules.tar` and `app.tar` with `/app` prefix.
+
+```starlark
+load("@aspect_rules_js//js:defs.bzl", "js_image_layer")
+
+js_image_layer(
+    name = "layers",
+    binary = "//label/to:js_binary",
+    root = "/app",
+)
+```
+"""
+
+load("@aspect_bazel_lib//lib:paths.bzl", "to_manifest_path")
+load("@bazel_skylib//lib:paths.bzl", "paths")
+
+_doc = """Create container image layers from js_binary targets.
+
+An example using rules_oci
+
+```starlark
+load("@aspect_rules_js//js:defs.bzl", "js_binary", "js_image_layer")
+load("@contrib_rules_oci//oci:defs.bzl", "oci_image")
+
+js_binary(
+    name = "binary",
+    entry_point = "main.js",
+)
+
+js_image_layer(
+    name = "layers",
+    binary = ":binary",
+    root = "/app"
+)
+
+oci_image(
+    name = "image",
+    tars = [
+        ":layers"
+    ]
+)
+```
+
+An example using legacy rules_docker
+
+See `e2e/js_image_rules_docker` for full example.
+
+```starlark
+load("@aspect_rules_js//js:defs.bzl", "js_binary", "js_image_layer")
+load("@io_bazel_rules_docker//container:container.bzl", "container_image")
+
+js_binary(
+    name = "main",
+    data = [
+        "//:node_modules/args-parser",
+    ],
+    entry_point = "main.js",
+)
+
+
+js_image_layer(
+    name = "layers",
+    binary = ":main",
+    root = "/app",
+    visibility = ["//visibility:__pkg__"],
+)
+
+filegroup(
+    name = "app_tar", 
+    srcs = [":layers"], 
+    output_group = "app"
+)
+container_layer(
+    name = "app_layer",
+    tars = [":app_tar"],
+)
+
+filegroup(
+    name = "node_modules_tar", 
+    srcs = [":layers"], 
+    output_group = "node_modules"
+)
+container_layer(
+    name = "node_modules_layer",
+    tars = [":node_modules_tar"],
+)
+
+container_image(
+    name = "image",
+    cmd = ["/app/main.sh"],
```

Comment:
nit: perhaps make an `/app/main` entry point that redirects to `main.sh` since the `.sh` extension is an implementation detail of `js_binary` and not otherwise visible in a BUILD file?

Reply:
I like the idea. though I am a little worried that it might make people neglect the fact we need bash to start the program.

Reply:
technically the extension should be `.bash` if it depends on bash 😉 

the bash dependency will be encoded in the generated executable file anyway:

```
#!/usr/bin/env bash
exec main.sh
```

Reply:
say no more 👍 done